pkg/syncer/upsync/upsync_reconcile.go

```diff
@@ -0,0 +1,261 @@
+	logger := klog.FromContext(ctx)
+
+	downstreamNamespace := ""
+	if upstreamNamespace != "" {
+		// retrieve the downstream namespace from the upstream namespace through the namespace locator index
+		namespaceLocator := shared.NewNamespaceLocator(upstreamClusterName, c.syncTargetClusterName, c.syncTargetUID, c.syncTargetName, upstreamNamespace)
+		locatorAnnotationValue, err := json.Marshal(namespaceLocator)
+		if err != nil {
+			return false, err
+		}
+		downstreamNamespaces, err := c.listDownstreamNamespacesByLocator(string(locatorAnnotationValue))
+		if err != nil {
+			return false, err
+		}
+		if len(downstreamNamespaces) == 1 {
+			namespace := downstreamNamespaces[0]
+			logger.WithValues(DownstreamName, namespace.GetName()).V(4).Info("Found downstream namespace for upstream namespace")
+			downstreamNamespace = namespace.GetName()
+		} else if len(downstreamNamespaces) > 1 {
+			// This should never happen unless there's some namespace collision.
+			var namespacesCollisions []string
+			for _, namespace := range downstreamNamespaces {
+				namespacesCollisions = append(namespacesCollisions, namespace.GetName())
+			}
+			return false, fmt.Errorf("(namespace collision) found multiple downstream namespaces: %s for upstream namespace %s|%s", strings.Join(namespacesCollisions, ","), upstreamClusterName, upstreamObject.GetNamespace())
+		} else {
+			logger.V(4).Info("No downstream namespaces found")
+			// Downstream resource namespace not present => delete resource upstream
+			return false, c.removeUpstreamResource(ctx, gvr, upstreamClusterName, upstreamNamespace, upstreamName, true)
+		}
+	}
+
+	logger = logger.WithValues(DownstreamNamespace, downstreamNamespace)
+	ctx = klog.NewContext(ctx, logger)
+
+	// retrieve downstream object
+
+	downstreamLister, err := c.getDownstreamLister(gvr)
+	if err != nil {
+		return false, err
+	}
+
+	var downstreamObject runtime.Object
+	if downstreamNamespace != "" {
+		downstreamObject, err = downstreamLister.ByNamespace(downstreamNamespace).Get(upstreamName)
+	} else {
+		downstreamObject, err = downstreamLister.Get(upstreamName)
+	}
+	if k8serror.IsNotFound(err) {
+		// Downstream resource not present => delete resource upstream
+		return false, c.removeUpstreamResource(ctx, gvr, upstreamClusterName, upstreamObject.GetNamespace(), upstreamObject.GetName(), true)
+	}
+	if err != nil {
+		return false, err
+	}
+
+	downstreamResource, ok := downstreamObject.(*unstructured.Unstructured)
+	if !ok {
+		return false, fmt.Errorf("type mismatch of resource object: received %T", downstreamResource)
+	}
+	resourceVersionDownstream := downstreamResource.GetResourceVersion()
+	markedForDeletionDownstream := downstreamResource.GetDeletionTimestamp() != nil
+
+	if upstreamObject == nil && !markedForDeletionDownstream {
+		// Resource doesn't exist upstream => let's create it upstream unless the corresponding downstream resource has been marked for deletion.
+		logger.Info("Creating resource upstream")
```

Comment:
V(1)

Reply:
Fixed in commit fffff6927c99c86702c7709881a77057331950cd